src/serializers/type_serializers/union.rs

```diff
@@ -326,36 +279,66 @@ impl TypeSerializer for TaggedUnionSerializer {
-                        let value_dict = value.downcast::<PyDict>().unwrap();
-                        lookup_key.py_get_dict_item(value_dict).ok()
-                    }
-                    false => lookup_key.simple_py_get_attr(value).ok(),
-                };
-                getattr_result.and_then(|opt| opt.map(|(_, bound)| bound.to_object(py)))
+                if let Ok(value_dict) = value.downcast::<PyDict>() {
+                    lookup_key.py_get_dict_item(value_dict).ok().flatten()
+                } else {
+                    lookup_key.simple_py_get_attr(value).ok().flatten()
+                }
+                .map(|(_, tag)| tag)
             }
-            Discriminator::Function(func) => func.call1(py, (value,)).ok(),
-        };
-        if discriminator_value.is_none() {
-            let value_str = truncate_safe_repr(value, None);
+            Discriminator::Function(func) => func.bind(py).call1((value,)).ok(),
+        }
+    }
 
-            // If extra.check is SerCheck::None, we're in a top-level union. We should thus raise this warning
-            if extra.check == SerCheck::None {
-                extra.warnings.custom_warning(
-                    format!(
-                        "Failed to get discriminator value for tagged union serialization with value `{value_str}` - defaulting to left to right union serialization."
-                    )
-                );
+    fn tagged_union_serialize<S>(
+        &self,
+        value: &Bound<'_, PyAny>,
+        // if this returns `Ok(v)`, we picked a union variant to serialize, where
+        // `S` is intermediate state which can be passed on to the finalizer
+        mut selector: impl FnMut(&CombinedSerializer, &Extra) -> PyResult<S>,
+        extra: &Extra,
+    ) -> PyResult<Option<S>> {
+        if let Some(tag) = self.get_discriminator_value(value) {
+            let mut new_extra = extra.clone();
+            new_extra.check = SerCheck::Strict;
+
+            let tag_str = tag.to_string();
+            if let Some(&serializer_index) = self.lookup.get(&tag_str) {
+                let selected_serializer = &self.choices[serializer_index];
+
+                match selector(selected_serializer, &new_extra) {
+                    Ok(v) => return Ok(Some(v)),
+                    Err(_) => {
+                        if self.retry_with_lax_check() {
+                            new_extra.check = SerCheck::Lax;
+                            if let Ok(v) = selector(selected_serializer, &new_extra) {
+                                return Ok(Some(v));
+                            }
+                        }
+                    }
+                }
             }
+        } else if extra.check == SerCheck::None {
+            // If extra.check is SerCheck::None, we're in a top-level union. We should thus raise
+            // this warning
+            let value_str = truncate_safe_repr(value, None);
+            extra.warnings.custom_warning(
```

Comment:
Moved this warning here from `get_discriminator_value`, seemed more correct as the other warnings are raised in `union_serialize`.